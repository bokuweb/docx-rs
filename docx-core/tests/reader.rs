use insta::assert_json_snapshot;

use docx_rs::*;
use std::fs::*;
use std::io::{Read, Write};

#[test]
pub fn read_hello() {
    let mut file = File::open("../fixtures/hello_world/hello_world.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_hello", &json);

    let path = std::path::Path::new("./tests/output/hello.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_numbering() {
    let mut file = File::open("../fixtures/numbering/numbering.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_numbering", &json);

    let path = std::path::Path::new("./tests/output/numbering.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_decoration() {
    let mut file = File::open("../fixtures/decoration/decoration.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_decoration", &json);

    let path = std::path::Path::new("./tests/output/decoration.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_highlight_and_underline() {
    let mut file =
        File::open("../fixtures/highlight_and_underline/highlight_and_underline.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_highlight_and_underline", &json);

    let path = std::path::Path::new("./tests/output/highlight_and_underline.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_history() {
    let mut file = File::open("../fixtures/history_libre_office/history.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_history", &json);

    let path = std::path::Path::new("./tests/output/history.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_indent_word_online() {
    let mut file = File::open("../fixtures/indent_word_online/indent.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_indent_word_online", &json);

    let path = std::path::Path::new("./tests/output/indent_word_online.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_tab_and_break() {
    let mut file = File::open("../fixtures/tab_and_break/tab_and_break.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_tab_and_break", &json);

    let path = std::path::Path::new("./tests/output/tab_and_break.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_table_docx() {
    let mut file = File::open("../fixtures/table_docx/table.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_table_docx", &json);

    let path = std::path::Path::new("./tests/output/table_docx.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_table_merged_libre_office() {
    let mut file = File::open("../fixtures/table_merged_libre_office/table_merged.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_table_merged_libre_office", &json);

    let path = std::path::Path::new("./tests/output/table_merged_libre_office.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_bom() {
    let mut file = File::open("../fixtures/bom/bom.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_bom", &json);

    let path = std::path::Path::new("./tests/output/bom.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_bookmark() {
    let mut file = File::open("../fixtures/bookmark/bookmark.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_bookmark", &json);

    let path = std::path::Path::new("./tests/output/bookmark.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_insert_table() {
    let mut file = File::open("../fixtures/insert_table/insert_table.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_insert_table", &json);

    let path = std::path::Path::new("./tests/output/insert_table.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_textbox() {
    let mut file = File::open("../fixtures/textbox/textbox.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_textbox", &json);

    let path = std::path::Path::new("./tests/output/textbox.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_from_doc() {
    let mut file = File::open("../fixtures/from_doc/from_doc.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_from_doc", &json);

    let path = std::path::Path::new("./tests/output/from_doc.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_lvl_override() {
    let mut file = File::open("../fixtures/lvl_override/override.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_lvl_override", &json);

    let path = std::path::Path::new("./tests/output/lvl_override.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_comment() {
    let mut file = File::open("../fixtures/comment/comment.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_comment", &json);

    let path = std::path::Path::new("./tests/output/comment.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_extended_comment() {
    let mut file = File::open("../fixtures/extended_comments/extended_comments.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("read_extended_comments", &json);

    let path = std::path::Path::new("./tests/output/extended_comments.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}

#[test]
pub fn read_line_spacing() {
    let mut file = File::open("../fixtures/line_spacing/line_spacing.docx").unwrap();
    let mut buf = vec![];
    file.read_to_end(&mut buf).unwrap();
    let json = read_docx(&buf).unwrap().json();

    assert_json_snapshot!("line_spacing", &json);

    let path = std::path::Path::new("./tests/output/line_spacing.json");
    let mut file = std::fs::File::create(&path).unwrap();
    file.write_all(json.as_bytes()).unwrap();
    file.flush().unwrap();
}
