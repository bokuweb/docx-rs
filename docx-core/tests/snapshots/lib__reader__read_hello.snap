---
source: docx-core/tests/reader.rs
expression: "&json"
---
"{\n  \"contentType\": {\n    \"types\": {\n      \"/_rels/.rels\": \"application/vnd.openxmlformats-package.relationships+xml\",\n      \"/docProps/app.xml\": \"application/vnd.openxmlformats-officedocument.extended-properties+xml\",\n      \"/docProps/core.xml\": \"application/vnd.openxmlformats-package.core-properties+xml\",\n      \"/word/_rels/document.xml.rels\": \"application/vnd.openxmlformats-package.relationships+xml\",\n      \"/word/comments.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.comments+xml\",\n      \"/word/document.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.document.main+xml\",\n      \"/word/fontTable.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.fontTable+xml\",\n      \"/word/numbering.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.numbering+xml\",\n      \"/word/settings.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.settings+xml\",\n      \"/word/styles.xml\": \"application/vnd.openxmlformats-officedocument.wordprocessingml.styles+xml\"\n    }\n  },\n  \"rels\": {\n    \"rels\": [\n      [\n        \"http://schemas.openxmlformats.org/package/2006/relationships/metadata/core-properties\",\n        \"rId1\",\n        \"docProps/core.xml\"\n      ],\n      [\n        \"http://schemas.openxmlformats.org/officeDocument/2006/relationships/extended-properties\",\n        \"rId2\",\n        \"docProps/app.xml\"\n      ],\n      [\n        \"http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument\",\n        \"rId3\",\n        \"word/document.xml\"\n      ]\n    ]\n  },\n  \"documentRels\": {\n    \"hasComments\": false,\n    \"hasNumberings\": false\n  },\n  \"docProps\": {\n    \"app\": {},\n    \"core\": {\n      \"config\": {\n        \"created\": null,\n        \"creator\": null,\n        \"description\": null,\n        \"language\": null,\n        \"lastModifiedBy\": null,\n        \"modified\": null,\n        \"revision\": null,\n        \"subject\": null,\n        \"title\": null\n      }\n    }\n  },\n  \"styles\": {\n    \"docDefaults\": {\n      \"runPropertyDefault\": {\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": null,\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        }\n      }\n    },\n    \"styles\": [\n      {\n        \"styleId\": \"Title\",\n        \"name\": \"Title\",\n        \"styleType\": \"paragraph\",\n        \"runProperty\": {\n          \"sz\": 56,\n          \"szCs\": 56,\n          \"color\": null,\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"Heading1\",\n        \"name\": \"Heading 1\",\n        \"styleType\": \"paragraph\",\n        \"runProperty\": {\n          \"sz\": 32,\n          \"szCs\": 32,\n          \"color\": \"2E74B5\",\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"Heading2\",\n        \"name\": \"Heading 2\",\n        \"styleType\": \"paragraph\",\n        \"runProperty\": {\n          \"sz\": 26,\n          \"szCs\": 26,\n          \"color\": \"2E74B5\",\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"Heading3\",\n        \"name\": \"Heading 3\",\n        \"styleType\": \"paragraph\",\n        \"runProperty\": {\n          \"sz\": 24,\n          \"szCs\": 24,\n          \"color\": \"1F4D78\",\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"Heading4\",\n        \"name\": \"Heading 4\",\n        \"styleType\": \"paragraph\",\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": \"2E74B5\",\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": true,\n          \"italicCs\": true,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"Heading5\",\n        \"name\": \"Heading 5\",\n        \"styleType\": \"paragraph\",\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": \"2E74B5\",\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"Heading6\",\n        \"name\": \"Heading 6\",\n        \"styleType\": \"paragraph\",\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": \"1F4D78\",\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"ListParagraph\",\n        \"name\": \"List Paragraph\",\n        \"styleType\": \"paragraph\",\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": null,\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"Hyperlink\",\n        \"name\": \"Hyperlink\",\n        \"styleType\": \"character\",\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": \"0563C1\",\n          \"highlight\": null,\n          \"underline\": \"single\",\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"FootnoteReference\",\n        \"name\": \"footnote reference\",\n        \"styleType\": \"character\",\n        \"runProperty\": {\n          \"sz\": null,\n          \"szCs\": null,\n          \"color\": null,\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"FootnoteText\",\n        \"name\": \"footnote text\",\n        \"styleType\": \"paragraph\",\n        \"runProperty\": {\n          \"sz\": 20,\n          \"szCs\": 20,\n          \"color\": null,\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      },\n      {\n        \"styleId\": \"FootnoteTextChar\",\n        \"name\": \"Footnote Text Char\",\n        \"styleType\": \"character\",\n        \"runProperty\": {\n          \"sz\": 20,\n          \"szCs\": 20,\n          \"color\": null,\n          \"highlight\": null,\n          \"underline\": null,\n          \"bold\": null,\n          \"boldCs\": null,\n          \"italic\": null,\n          \"italicCs\": null,\n          \"vanish\": null\n        },\n        \"paragraphProperty\": {\n          \"runProperty\": {\n            \"sz\": null,\n            \"szCs\": null,\n            \"color\": null,\n            \"highlight\": null,\n            \"underline\": null,\n            \"bold\": null,\n            \"boldCs\": null,\n            \"italic\": null,\n            \"italicCs\": null,\n            \"vanish\": null\n          },\n          \"style\": \"Normal\",\n          \"numberingProperty\": null,\n          \"alignment\": null,\n          \"indent\": null\n        }\n      }\n    ]\n  },\n  \"document\": {\n    \"children\": [\n      {\n        \"type\": \"paragraph\",\n        \"data\": {\n          \"children\": [\n            {\n              \"type\": \"run\",\n              \"data\": {\n                \"runProperty\": {\n                  \"sz\": null,\n                  \"szCs\": null,\n                  \"color\": null,\n                  \"highlight\": null,\n                  \"underline\": null,\n                  \"bold\": null,\n                  \"boldCs\": null,\n                  \"italic\": null,\n                  \"italicCs\": null,\n                  \"vanish\": null\n                },\n                \"children\": [\n                  {\n                    \"type\": \"text\",\n                    \"data\": {\n                      \"preserveSpace\": true,\n                      \"text\": \"Hello World\"\n                    }\n                  }\n                ]\n              }\n            }\n          ],\n          \"property\": {\n            \"runProperty\": {\n              \"sz\": null,\n              \"szCs\": null,\n              \"color\": null,\n              \"highlight\": null,\n              \"underline\": null,\n              \"bold\": null,\n              \"boldCs\": null,\n              \"italic\": null,\n              \"italicCs\": null,\n              \"vanish\": null\n            },\n            \"style\": \"Normal\",\n            \"numberingProperty\": null,\n            \"alignment\": null,\n            \"indent\": null\n          },\n          \"hasNumbering\": false,\n          \"attrs\": []\n        }\n      }\n    ],\n    \"sectionProperty\": {\n      \"pageSize\": {\n        \"w\": 11906,\n        \"h\": 16838\n      },\n      \"pageMargin\": {\n        \"top\": 1985,\n        \"left\": 1701,\n        \"bottom\": 1701,\n        \"right\": 1701,\n        \"header\": 851,\n        \"footer\": 992,\n        \"gutter\": 0\n      },\n      \"columns\": 425,\n      \"documentGrid\": 360\n    },\n    \"hasNumbering\": false\n  },\n  \"comments\": {\n    \"comments\": []\n  },\n  \"numberings\": {\n    \"abstractNums\": [\n      {\n        \"id\": 0,\n        \"styleLink\": null,\n        \"numStyleLink\": null,\n        \"levels\": [\n          {\n            \"level\": 0,\n            \"start\": 1,\n            \"format\": \"bullet\",\n            \"text\": \"●\",\n            \"jc\": \"left\",\n            \"pstyle\": null,\n            \"paragraphProperty\": {\n              \"runProperty\": {\n                \"sz\": null,\n                \"szCs\": null,\n                \"color\": null,\n                \"highlight\": null,\n                \"underline\": null,\n                \"bold\": null,\n                \"boldCs\": null,\n                \"italic\": null,\n                \"italicCs\": null,\n                \"vanish\": null\n              },\n              \"style\": \"Normal\",\n              \"numberingProperty\": null,\n              \"alignment\": null,\n              \"indent\": {\n                \"start\": 720,\n                \"startChars\": null,\n                \"end\": null,\n                \"specialIndent\": {\n                  \"type\": \"hanging\",\n                  \"val\": 360\n                }\n              }\n            }\n          },\n          {\n            \"level\": 1,\n            \"start\": 1,\n            \"format\": \"bullet\",\n            \"text\": \"○\",\n            \"jc\": \"left\",\n            \"pstyle\": null,\n            \"paragraphProperty\": {\n              \"runProperty\": {\n                \"sz\": null,\n                \"szCs\": null,\n                \"color\": null,\n                \"highlight\": null,\n                \"underline\": null,\n                \"bold\": null,\n                \"boldCs\": null,\n                \"italic\": null,\n                \"italicCs\": null,\n                \"vanish\": null\n              },\n              \"style\": \"Normal\",\n              \"numberingProperty\": null,\n              \"alignment\": null,\n              \"indent\": {\n                \"start\": 1440,\n                \"startChars\": null,\n                \"end\": null,\n                \"specialIndent\": {\n                  \"type\": \"hanging\",\n                  \"val\": 360\n                }\n              }\n            }\n          },\n          {\n            \"level\": 2,\n            \"start\": 1,\n            \"format\": \"bullet\",\n            \"text\": \"■\",\n            \"jc\": \"left\",\n            \"pstyle\": null,\n            \"paragraphProperty\": {\n              \"runProperty\": {\n                \"sz\": null,\n                \"szCs\": null,\n                \"color\": null,\n                \"highlight\": null,\n                \"underline\": null,\n                \"bold\": null,\n                \"boldCs\": null,\n                \"italic\": null,\n                \"italicCs\": null,\n                \"vanish\": null\n              },\n              \"style\": \"Normal\",\n              \"numberingProperty\": null,\n              \"alignment\": null,\n              \"indent\": {\n                \"start\": 2160,\n                \"startChars\": null,\n                \"end\": null,\n                \"specialIndent\": {\n                  \"type\": \"hanging\",\n                  \"val\": 360\n                }\n              }\n            }\n          },\n          {\n            \"level\": 3,\n            \"start\": 1,\n            \"format\": \"bullet\",\n            \"text\": \"●\",\n            \"jc\": \"left\",\n            \"pstyle\": null,\n            \"paragraphProperty\": {\n              \"runProperty\": {\n                \"sz\": null,\n                \"szCs\": null,\n                \"color\": null,\n                \"highlight\": null,\n                \"underline\": null,\n                \"bold\": null,\n                \"boldCs\": null,\n                \"italic\": null,\n                \"italicCs\": null,\n                \"vanish\": null\n              },\n              \"style\": \"Normal\",\n              \"numberingProperty\": null,\n              \"alignment\": null,\n              \"indent\": {\n                \"start\": 2880,\n                \"startChars\": null,\n                \"end\": null,\n                \"specialIndent\": {\n                  \"type\": \"hanging\",\n                  \"val\": 360\n                }\n              }\n            }\n          },\n          {\n            \"level\": 4,\n            \"start\": 1,\n            \"format\": \"bullet\",\n            \"text\": \"○\",\n            \"jc\": \"left\",\n            \"pstyle\": null,\n            \"paragraphProperty\": {\n              \"runProperty\": {\n                \"sz\": null,\n                \"szCs\": null,\n                \"color\": null,\n                \"highlight\": null,\n                \"underline\": null,\n                \"bold\": null,\n                \"boldCs\": null,\n                \"italic\": null,\n                \"italicCs\": null,\n                \"vanish\": null\n              },\n              \"style\": \"Normal\",\n              \"numberingProperty\": null,\n              \"alignment\": null,\n              \"indent\": {\n                \"start\": 3600,\n                \"startChars\": null,\n                \"end\": null,\n                \"specialIndent\": {\n                  \"type\": \"hanging\",\n                  \"val\": 360\n                }\n              }\n            }\n          },\n          {\n            \"level\": 5,\n            \"start\": 1,\n            \"format\": \"bullet\",\n            \"text\": \"■\",\n            \"jc\": \"left\",\n            \"pstyle\": null,\n            \"paragraphProperty\": {\n              \"runProperty\": {\n                \"sz\": null,\n                \"szCs\": null,\n                \"color\": null,\n                \"highlight\": null,\n                \"underline\": null,\n                \"bold\": null,\n                \"boldCs\": null,\n                \"italic\": null,\n                \"italicCs\": null,\n                \"vanish\": null\n              },\n              \"style\": \"Normal\",\n              \"numberingProperty\": null,\n              \"alignment\": null,\n              \"indent\": {\n                \"start\": 4320,\n                \"startChars\": null,\n                \"end\": null,\n                \"specialIndent\": {\n                  \"type\": \"hanging\",\n                  \"val\": 360\n                }\n              }\n            }\n          },\n          {\n            \"level\": 6,\n            \"start\": 1,\n            \"format\": \"bullet\",\n            \"text\": \"●\",\n            \"jc\": \"left\",\n            \"pstyle\": null,\n            \"paragraphProperty\": {\n              \"runProperty\": {\n                \"sz\": null,\n                \"szCs\": null,\n                \"color\": null,\n                \"highlight\": null,\n                \"underline\": null,\n                \"bold\": null,\n                \"boldCs\": null,\n                \"italic\": null,\n                \"italicCs\": null,\n                \"vanish\": null\n              },\n              \"style\": \"Normal\",\n              \"numberingProperty\": null,\n              \"alignment\": null,\n              \"indent\": {\n                \"start\": 5040,\n                \"startChars\": null,\n                \"end\": null,\n                \"specialIndent\": {\n                  \"type\": \"hanging\",\n                  \"val\": 360\n                }\n              }\n            }\n          },\n          {\n            \"level\": 7,\n            \"start\": 1,\n            \"format\": \"bullet\",\n            \"text\": \"○\",\n            \"jc\": \"left\",\n            \"pstyle\": null,\n            \"paragraphProperty\": {\n              \"runProperty\": {\n                \"sz\": null,\n                \"szCs\": null,\n                \"color\": null,\n                \"highlight\": null,\n                \"underline\": null,\n                \"bold\": null,\n                \"boldCs\": null,\n                \"italic\": null,\n                \"italicCs\": null,\n                \"vanish\": null\n              },\n              \"style\": \"Normal\",\n              \"numberingProperty\": null,\n              \"alignment\": null,\n              \"indent\": {\n                \"start\": 5760,\n                \"startChars\": null,\n                \"end\": null,\n                \"specialIndent\": {\n                  \"type\": \"hanging\",\n                  \"val\": 360\n                }\n              }\n            }\n          },\n          {\n            \"level\": 8,\n            \"start\": 1,\n            \"format\": \"bullet\",\n            \"text\": \"■\",\n            \"jc\": \"left\",\n            \"pstyle\": null,\n            \"paragraphProperty\": {\n              \"runProperty\": {\n                \"sz\": null,\n                \"szCs\": null,\n                \"color\": null,\n                \"highlight\": null,\n                \"underline\": null,\n                \"bold\": null,\n                \"boldCs\": null,\n                \"italic\": null,\n                \"italicCs\": null,\n                \"vanish\": null\n              },\n              \"style\": \"Normal\",\n              \"numberingProperty\": null,\n              \"alignment\": null,\n              \"indent\": {\n                \"start\": 6480,\n                \"startChars\": null,\n                \"end\": null,\n                \"specialIndent\": {\n                  \"type\": \"hanging\",\n                  \"val\": 360\n                }\n              }\n            }\n          }\n        ]\n      }\n    ],\n    \"numberings\": [\n      {\n        \"id\": 1,\n        \"abstractNumId\": 0\n      }\n    ]\n  },\n  \"settings\": {\n    \"defaultTabStop\": 709,\n    \"zoom\": 100\n  },\n  \"fontTable\": {}\n}"
